### ISSUE TITLE
synchronize with USB disk

### ISSUE THREAD

**Issue by cycomanic** [AmesCornish/buttersink#1]
Hi,

I was under the impression that it should also be possible to synchronize snapshots between local btrfs filesystems, however when I try this I get an error, which seems to be due to the fact that both the source and destination have the same rootid. 

$sudo ./buttersink.py -n btrfs:///home/.snapshot/ btrfs:///media/BackupDisk1/ -l log.txt 

```
Traceback (most recent call last):
  File "./buttersink.py", line 168, in main
    dest = parseSink(args.dest, source is not None, args.dry_run)
  File "./buttersink.py", line 152, in parseSink
    return Sinks[parts['method']](parts['host'], parts['path'], isDest, dryrun)
  File "/home/jschrod/Downloads/System/BackupProgs/buttersink/buttersink/ButterStore.py", line 53, in __init__
    self._fillVolumesAndPaths()
  File "/home/jschrod/Downloads/System/BackupProgs/buttersink/buttersink/ButterStore.py", line 71, in _fillVolumesAndPaths
    for bv in mount.subvolumes:
  File "/home/jschrod/Downloads/System/BackupProgs/buttersink/buttersink/btrfs.py", line 485, in     subvolumes
    self._getRoots()
  File "/home/jschrod/Downloads/System/BackupProgs/buttersink/buttersink/btrfs.py", line 647, in _getRoots
    info,
  File "/home/jschrod/Downloads/System/BackupProgs/buttersink/buttersink/btrfs.py", line 371, in __init__
    assert rootid not in Volume.volumes, rootid
AssertionError: 5
```

here's the last lines from the log file:

```
2015-02-08 11:22:06,452:   DEBUG:btrfs.py[552] _getMounts(): /: /media/BackupDisk1
2015-02-08 11:22:06,452:   DEBUG:btrfs.py[595] _walkTree(): Reading 14 nodes from 3992 bytes
2015-02-08 11:22:06,453:   DEBUG:btrfs.py[352] __init__(): Volume 5/0:    StructureTuple(inode=StructureTuple(generation=1, transid=0, size=3, nbytes=16384, block_group=0, nlink=1, uid=0, gid=0, mode=16877, rdev=0, flags=18446744071562067968L, sequence=0, reserved='', atime=StructureTuple(sec=0, nsec=0), ctime=StructureTuple(sec=0, nsec=0), mtime=StructureTuple(sec=0, nsec=0), otime=StructureTuple(sec=0, nsec=0)), generation=28, root_dirid=256, bytenr=31227904, byte_limit=0, bytes_used=16384, last_snapshot=0, flags=0, refs=1, drop_progress=StructureTuple(objectid=0, type=0, offset=0), drop_level=0, level=0, generation_v2=28, uuid=None, parent_uuid=None, received_uuid=None, ctransid=28, otransid=0, stransid=0, rtransid=0, ctime=StructureTuple(sec=1423352843, nsec=480916699), otime=StructureTuple(sec=0, nsec=0), stime=StructureTuple(sec=0, nsec=0), rtime=StructureTuple(sec=0, nsec=0), reserved='')
2015-02-08 11:22:06,453:   ERROR:buttersink.py[222] main(): 
Traceback (most recent call last):
  File "./buttersink.py", line 168, in main
    dest = parseSink(args.dest, source is not None, args.dry_run)
  File "./buttersink.py", line 152, in parseSink
    return Sinks[parts['method']](parts['host'], parts['path'], isDest, dryrun)
  File "/home/jschrod/Downloads/System/BackupProgs/buttersink/buttersink/ButterStore.py", line 53, in __init__
    self._fillVolumesAndPaths()
  File "/home/jschrod/Downloads/System/BackupProgs/buttersink/buttersink/ButterStore.py", line 71, in _fillVolumesAndPaths
    for bv in mount.subvolumes:
  File "/home/jschrod/Downloads/System/BackupProgs/buttersink/buttersink/btrfs.py", line 485, in subvolumes
    self._getRoots()
  File "/home/jschrod/Downloads/System/BackupProgs/buttersink/buttersink/btrfs.py", line 647, in _getRoots
    info,
  File "/home/jschrod/Downloads/System/BackupProgs/buttersink/buttersink/btrfs.py", line 371, in  __init__
    assert rootid not in Volume.volumes, rootid
AssertionError: 5                    
```

$sudo btrfs subvolume list /media/BackupDisk1/ 

```
ID 257 gen 18 top level 5 path @ButtersinkBackups
ID 258 gen 19 top level 257 path @ButtersinkBackups/@home
ID 259 gen 11 top level 257 path @ButtersinkBackups/@jschrod-Pictures
ID 260 gen 12 top level 257 path @ButtersinkBackups/@jschrod-Uni
ID 261 gen 13 top level 257 path @ButtersinkBackups/@jschrod-Misc 
ID 262 gen 14 top level 257 path @ButtersinkBackups/@jschrod-Documents
ID 263 gen 15 top level 257 path @ButtersinkBackups/@chris-Pictures
ID 264 gen 16 top level 257 path @ButtersinkBackups/@chris-Documents
ID 265 gen 17 top level 257 path @ButtersinkBackups/@chris-Desktop
ID 266 gen 18 top level 257 path @ButtersinkBackups/@chris-Downloads
ID 267 gen 28 top level 5 path @home
```

$sudo btrfs subvolume list /home/

```
ID 325 gen 1996 top level 5 path .snapshot/monthly_2015-01-15_12:30:40
ID 341 gen 2720 top level 5 path .snapshot/weekly_2015-01-18_06:57:16
ID 425 gen 8387 top level 5 path .snapshot/weekly_2015-01-25_08:46:03
ID 529 gen 11064 top level 5 path .snapshot/weekly_2015-02-01_09:45:57
ID 539 gen 11567 top level 5 path .snapshot/daily_2015-02-02_10:19:18
ID 548 gen 12033 top level 5 path .snapshot/daily_2015-02-03_09:22:20
ID 555 gen 12549 top level 5 path .snapshot/daily_2015-02-04_09:56:22
ID 574 gen 13593 top level 5 path .snapshot/daily_2015-02-05_11:59:55
ID 586 gen 14452 top level 5 path .snapshot/daily_2015-02-06_12:16:58
ID 595 gen 14784 top level 5 path .snapshot/daily_2015-02-07_06:53:20
ID 597 gen 14896 top level 5 path .snapshot/hourly_2015-02-07_08:17:01
ID 598 gen 14986 top level 5 path .snapshot/hourly_2015-02-07_09:17:01
ID 599 gen 15071 top level 5 path .snapshot/hourly_2015-02-07_17:17:01
ID 600 gen 15073 top level 5 path .snapshot/hourly_2015-02-07_18:17:01
ID 601 gen 15076 top level 5 path .snapshot/hourly_2015-02-07_19:17:01
ID 602 gen 15079 top level 5 path .snapshot/hourly_2015-02-07_20:17:01
ID 603 gen 15110 top level 5 path .snapshot/hourly_2015-02-07_21:17:01
ID 604 gen 15151 top level 5 path .snapshot/hourly_2015-02-07_22:17:01
ID 605 gen 15168 top level 5 path .snapshot/daily_2015-02-08_06:23:32
ID 606 gen 15177 top level 5 path .snapshot/weekly_2015-02-08_06:28:51
ID 607 gen 15259 top level 5 path .snapshot/hourly_2015-02-08_07:17:01
ID 608 gen 15296 top level 5 path .snapshot/hourly_2015-02-08_09:17:01
ID 609 gen 15336 top level 5 path .snapshot/hourly_2015-02-08_10:17:02
ID 610 gen 15406 top level 5 path .snapshot/hourly_2015-02-08_11:17:01
```

Am I doing something wrong?

**Comment by AmesCornish** [comment]
Cyco,

No, you're not doing anything wrong.  I have not yet implemented the
backend to sync with another btrfs filesystem.  I'm personally just using
it for backups to and restores from S3.  It's not that difficult a feature,
however, and I'll boost it up my priority list because of your interest.

I'm not positive that the error you're seeing is just because of your sync
destination.  Have you been able to buttersink up to S3?

Thanks for the detailed bug report!
- Ames

## 

Ames Cornish  ~  http://montebellopartners.com/
650-533-0835  ~  ames@montebellopartners.com

**Comment by zetasyanthis** [comment]
Adding a somewhat-related query to this.  Does that response mean the listed SSH functionality is also not currently present?

**Comment by cycomanic** [comment]
Hi Ames,

thanks that clears things up. I have not used S3 with buttersink. I might try when I have some more time, however the ability to do local syncs would be great.

Cheers
Jochen

**Comment by AmesCornish** [comment]
I've just released a new version of buttersink, 0.6, that has support for
syncing with a remote btrfs over ssh and with a second local btrfs.  If you
try it, please let me know how it works for you.

Thanks!
- Ames

## 

Ames Cornish  ~  http://montebellopartners.com/
650-533-0835  ~  ames@montebellopartners.com
